R/processCapabilityStudies.R

```diff
@@ -270,4 +271,96 @@ processCapabilityStudies <- function(jaspResults, dataset, options){
+    
+    tau <- standardDeviation / average
+    sdLog <- sqrt(log(tau^2 +1))
+    meanLog <- log(average) - ((sdLog^2) / 2)
+    partsLower <- plnorm(q = LSL, meanlog = meanLog, sdlog = sdLog)
+    partsUpper <- 1- plnorm(q = USL,meanlog = meanLog, sdlog = sdLog)
+    CPK <- 1- plnorm((max(partsLower,partsUpper)))/3
+    
+    nonNormalCapabilityTable$addRows(list("average" = average,
+                                          "standardDeviation" = standardDeviation,
+                                          "LSL" = LSL,
+                                          "USL" = USL,
+                                          "CPK" = CPK,
+                                          "partsLower" = partsLower,
+                                          "partsUpper" = partsUpper))
+  }
+  
+  if (dis == "Weibull"){
+    nonNormalCapabilityTable <- createJaspTable(title = gettext("Process Capability based on Weibull distribution"))
+    nonNormalCapabilityTable$dependOn(c("upperSpecification","lowerSpecification","targetValue","diameter","subgroups"))
+    
+    nonNormalCapabilityTable$addColumnInfo(name = "average", type = "number", title = gettext("Average"))
+    nonNormalCapabilityTable$addColumnInfo(name = "standardDeviation", type = "number", title = gettext("Standard Deviation"))
+    nonNormalCapabilityTable$addColumnInfo(name = "LSL", type = "number", title = gettext("LSL"))
+    nonNormalCapabilityTable$addColumnInfo(name = "USL", type = "number", title = gettext("USL"))
+    nonNormalCapabilityTable$addColumnInfo(name = "beta", type = "number", title = gettext("Beta"))
+    nonNormalCapabilityTable$addColumnInfo(name = "theta", type = "number", title = gettext("Theta"))
+    nonNormalCapabilityTable$addColumnInfo(name = "CPK", type = "number", title = gettext("Cpk"))
+    
+    dataDiameter <- .convertDatasetToQccReady(dataset, diameter, subgroupsName)
+    
+    USL <- as.numeric(options$upperSpecification)
+    LSL <- as.numeric(options$lowerSpecification)
+    targetValue <- as.numeric(options$targetValue)
+    sampleMean <- mean(as.matrix(dataDiameter),na.rm = TRUE)
+    
+    q <- qcc::qcc(dataDiameter, type ='S', plot=FALSE)
+
+    stDevOverall <- sd(as.matrix(dataDiameter),na.rm = TRUE)
+    standardDeviation <- stDevOverall
+    average <- sampleMean
+    beta <- mixdist::weibullpar(mu = average, sigma = standardDeviation, loc = 0)$shape
+    theta <- mixdist::weibullpar(mu = average, sigma = standardDeviation, loc = 0)$scale
+    CPK <- 0
+    
+    nonNormalCapabilityTable$addRows(list("average" = average,
+                                          "standardDeviation" = standardDeviation,
+                                          "LSL" = LSL,
+                                          "USL" = USL,
+                                          "CPK" = CPK,
+                                          "beta" = beta,
+                                          "theta" = theta))
+  }
```

Comment:
so everything that these if statements have in common, should not be copied in each if statements.

For example:
```r
dataDiameter <- .convertDatasetToQccReady(dataset, diameter, subgroupsName)

USL <- as.numeric(options$upperSpecification)
LSL <- as.numeric(options$lowerSpecification)
targetValue <- as.numeric(options$targetValue)
sampleMean <- mean(as.matrix(dataDiameter),na.rm = TRUE)

q <- qcc::qcc(dataDiameter, type ='S', plot=FALSE)

if (dis == "Lognormal") {
  
  nonNormalCapabilityTable <- createJaspTable(title = gettext("Process Capability based on Lognormal distribution"))
  nonNormalCapabilityTable$addColumnInfo(...)
  # more of addColumnInfo

  stDevWithin <- q$std.dev   #stDevWithin <- rBar/d2
  stDevOverall <- sd(as.matrix(dataDiameter),na.rm = TRUE)

} else if (dis == "Weibull") {

  nonNormalCapabilityTable <- createJaspTable(title = gettext("Process Capability based on Weibull distribution"))
  nonNormalCapabilityTable$addColumnInfo(...)
  # more of addColumnInfo

  stDevOverall <- sd(as.matrix(dataDiameter),na.rm = TRUE)
    standardDeviation <- stDevOverall

}

nonNormalCapabilityTable$dependOn(c("upperSpecification","lowerSpecification","targetValue","diameter","subgroups"))

return(nonNormalCapabilityTable)

```